Best practices document

asmeurer opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

This is a new document in the docs that outlines some of the best practices for using SymPy.

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes https://github.com/sympy/sympy/issues/24597

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- other
  - Add a "best practices" page to the documentation.
<!-- END RELEASE NOTES -->

asmeurer committed:
Start best practices document

asmeurer committed:
Write section on avoiding using strings as input

asmeurer committed:
Cleanup some unwritten section names

asmeurer committed:
Add some more to-write sections

asmeurer committed:
Add some other sections for best practices

asmeurer committed:
Write a best practice section for defining symbols

asmeurer committed:
Fix doctests

asmeurer committed:
Start writing best practice for avoiding floats

asmeurer committed:
Add best practices section on avoiding simplify()

asmeurer committed:
Add best practices section on not hard-coding symbol names

asmeurer committed:
Write some more on the avoiding floats section

I'm still not happy with the way this section looks.

asmeurer committed:
Merge branch 'master' into best-practices

asmeurer committed:
Add best practices section on args invariant

asmeurer committed:
Add a best practice about comparing symbolic objects

asmeurer committed:
Add a best practice for assumptions handlers

asmeurer committed:
Add a best practices section on avoiding too much automatic evaluation

asmeurer committed:
Add TODOs for missing sections

asmeurer committed:
Write best practices sections on storing extra data and redefining __eq__

asmeurer committed:
Fix doctest failures

asmeurer committed:
Write best practices section on not denesting collections

asmeurer committed:
Some improvements to the best practices document

asmeurer committed:
Merge branch 'master' into best-practices

asmeurer committed:
Fix a doctest failure

asmeurer committed:
Some text improvements to the best practices document

asmeurer committed:
Write the symbolic vs. numeric section of the best practices document

asmeurer committed:
Only doctest the best practices document when numpy is installed

asmeurer committed:
Remove the "interactive vs. programmatic" section

asmeurer committed:
Note that the best practices page is primarily for general best practices

asmeurer committed:
Various text updates in the best practices document

asmeurer committed:
More text updates to the best practices doc

asmeurer committed:
Better header name

asmeurer committed:
Cleanups to best practices document

asmeurer committed:
Updates to the best practices document

asmeurer committed:
Merge branch 'master' into best-practices

asmeurer commented on doc/src/explanation/best-practices.md:
```diff
@@ -0,0 +1,1329 @@
+simplification using {meth}`~sympy.core.basic.Basic.replace`, or in general, manually using
+[advanced expression manipulation](tutorial-manipulation).
+
+<!-- TODO: Add an example -->
```
Can anyone suggest a good example to demonstrate why using `simplify()` should be avoided (especially in programmatic contexts) over more targeted simplification functions like `cancel()`?

smichr replied:
I think @oscarbenjamin mentioned one recently that had to do with hyperbolics appearing that made some roots more complicated. Trig functions without targeted (Fu-like) transpormation are notorious for getting unwieldy. Also, even something simple like 
```python
>>> exp(x)/2 + exp(-x)/2
exp(x)/2 + exp(-x)/2
>>> simplify(_)
cosh(x)
```
is simpler but one might not have wanted to convert to hyperbolics.

oscarbenjamin replied:
The example from https://github.com/sympy/sympy/pull/25458#issuecomment-1662397968 is a good one:
```python
In [10]: M = Matrix([[0, -t*cos(t) + sin(t)], [-t*cos(t) + sin(t), 0]])

In [11]: M
Out[11]: 
⎡        0           -t⋅cos(t) + sin(t)⎤
⎢                                      ⎥
⎣-t⋅cos(t) + sin(t)          0         ⎦

In [12]: p = M.charpoly(z).as_expr()

In [13]: p
Out[13]: 
 2                      2
z  - (t⋅cos(t) - sin(t)) 

In [14]: roots(p, z)
Out[14]: 
⎧    ___________________________________        ___________________________________   ⎫
⎨   ╱  2    2                      2           ╱  2    2                      2       ⎬
⎩-╲╱  t ⋅cos (t) - t⋅sin(2⋅t) + sin (t) : 1, ╲╱  t ⋅cos (t) - t⋅sin(2⋅t) + sin (t) : 1⎭
```
Why didn't roots spot that there is a simpler expression for the roots here? Where did `sin(2*t)` come from?

The answer is:
```python
In [15]: p.expand()
Out[15]: 
   2    2                           2      2   
- t ⋅cos (t) + 2⋅t⋅sin(t)⋅cos(t) + z  - sin (t)

In [16]: p.expand().simplify()
Out[16]: 
   2    2                    2      2   
- t ⋅cos (t) + t⋅sin(2⋅t) + z  - sin (t)

In [17]: p.expand().collect(z, factor)
Out[17]: 
 2                      2
z  - (t⋅cos(t) - sin(t)) 

In [18]: p.expand().simplify().collect(z, factor)
Out[18]: 
   2    2                    2      2   
- t ⋅cos (t) + t⋅sin(2⋅t) + z  - sin (t)
```
So somewhere `simplify` (or `trigsimp`) is used and it decides to replace `2*sin(t)*cos(t)` with `sin(2*t)`. Whether or not this is a good idea is very context dependent but if you intend to do algebra like `factor`, `cancel` etc then it is definitely not better to introduce new "generators" (although eliminating some generators in favour of others that are *already* present is usually good). Here the use of `simplify` turns an expression that could have been factorised into one that cannot be factorised. In turn this takes us from nice rational expressions for the roots that are free of branch cuts to horrible trig-radical expressions.

I'm not sure why `eigenvals` itself is not affected but all of these functions misuse or overuse `simplify`:
```python
In [20]: M.eigenvals()
Out[20]: {-t⋅cos(t) + sin(t): 1, t⋅cos(t) - sin(t): 1}
```
The fact that `simplify` is actually being used *inside* functions like `roots`, `eigenvals` etc means that it is not even possible to carefully manipulate your inputs into the form you want before calling those functions.

asmeurer replied:
I'll have to think about this. A potential problem with an example like this is that it's hard to not just say "well, shouldn't `simplify()` just be smarter?" (e.g., in this case, try to rewrite the expression in such a way that makes it more factorable before calling `factor()`). And indeed, it's quite possible that at some point in the future it *will* become smarter, which would completely break the example.

It would be more robust to have an example where simplify() pretty unambiguously returns a "simple" form of an expression by some measure, but you can also say by some other measure that it isn't simple.

Maybe I can manipulate this example into something like that, but it's hard because you're really just saying "`simplify()` returned an expression that wasn't the correct 'simplified' for `simplify()`". But it would be I think more instructive to show an example like "`simplify()` returned something that wasn't the correct 'simplified' for `<some other function>`".

Another idea: even just for a basic rational function, there's at least 3 "simple" ways it can be represented: as p/q with p and q expanded and no factors in common (i.e., `cancel()`), with p/q with p and q factored (`factor()`), or as a partial fraction decomposition (`apart()`). `simplify()` seems to prefer the first of these, but there's no reason in particular for that, and it might some day switch to one of the others. But the point is that there are situations where all three forms can be useful.

oscarbenjamin replied:
I think that anything that you might write here would be potentially subject to change in future because simplify is heuristic. The point of my example is that it is a real problem that is actually caused by the use of simplify right now in the SymPy codebase. Of course we *should* try to fix that in future but it demonstrates why naive use of simplify is not always good (besides speed concerns) because it can lead to unpredictable results that ruin a larger calculation.

sylee957 replied:
I think that the problem of simplify

- Ambiguity problem:
  In the PR https://github.com/sympy/sympy/pull/25456, it is discovered that simplify is very hard to trigger and is very difficult to quantify when does it trigger.

- Abstraction problem:
  In the dependency inversion principle, high level features shouldn't depend on the low level implementation, but abstractions of it.
 `simplify` is very specific to the implementation details of computer algebra itself, than the abstract algebra. For example, I'm very skeptical about if transfer function should compute feedback result, or quaternions should compute norms with `simplify` than do bare algebra.
  This can later cause significant restriction if you want to let users of use implementations of symengine, or do other customizations.

sylee957 replied:
I think that the only production and practical usage of `simplify` should be for compilation workflow.
And similar as how compiler works, the compiled code can never be reused by humans, and can only interpreted by compilers or machines.
Similar as that, the result of simplify is never interpretable than other `simplify` or the printers.

oscarbenjamin replied:
I don't think that `simplify` should even be used for a compilation workflow: if you want to optimise an expression for turning into compiled code then you want a specialised simplify function for that.

Basically I don't think there is any situation within SymPy itself where `simplify` should be used but it is a nice thing for users to use interactively that will likely do what they want in simple cases.

asmeurer replied:
That's actually a good point. I think in practice, more people have the issue that simplify does too little, that it doesn't happen to actually call the simplification function they are looking for.

asmeurer commented on doc/src/explanation/best-practices.md:
```diff
@@ -0,0 +1,1329 @@
+the `args` of expressions directly (see the [custom functions
+guide](custom-functions-assumptions) for an example).
+
+<!-- TODO: Give an example here. Can we show an infinite recursion? -->
```
@oscarbenjamin do you know of an example of when assumptions handles can actually lead to infinite recursion? I couldn't come up with one easily, but I think maybe the problem is that this mainly happens inside of core SymPy classes rather than custom classes because those are the ones that are more likely to call each other.

oscarbenjamin replied:
> do you know of an example of when assumptions handles can actually lead to infinite recursion?

You have an example below. Do you mean something from the codebase though?

Generally I would remove any code that does this from the codebase.

asmeurer replied:
I mean an illustrative example I can include here. The example below is based on infinite recursion from evaluating assumptions on `self`, but I am looking for an example based on the "don't create new expressions in assumptions handlers" rule.

oscarbenjamin replied:
I can contrive an example but as you say it mostly happens in core class e.g. because someone does not consider that `1 + self` creates a new `Add` or something like that.

Here are examples of what should not be done:
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L618
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L661-L664
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L680
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L746
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L748-L750
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L754-L756

There are many more in the same file. Maybe you can identify an immediate bug from those. There are two types of infinite recursion that can occur:

- Recursing onto a different object that then recurses onto another different object and so on.
- Recursing onto a larger object that contains the current object so that the query comes back down to the same query on the original object.

It is very difficult to predict when any created expressions might lead to either of these cases because automatic evaluation rules can be so unpredictable. I don't see how else to define a rule that could control it besides "never create a new expression".

asmeurer replied:
To be fair, some of those seem hard to do without just completely copying the assumptions logic from the other class. I guess I basically already said the same thing in the other issue, but this really is a fundamental limitation of this recursive way of implementing assumptions handlers, which makes them responsible for both returning the result of the assumptions query and, to some extent, the method by which other queries are logically deduced. A better design is to separate out the order that the queries are actually made in to a top-level solver, and represent facts in an symbolic way.

Anyway, I think this is probably hard to show an explicit example for. Maybe there's an issue or PR I could reference where something like this happened. Otherwise, I'm just going to leave this as it is and remove the TODO.

smichr commented on doc/src/explanation/best-practices.md:
```diff
@@ -0,0 +1,1329 @@
+### Defining Symbols
+
+- **The best way to define symbols is using the {func}`~.symbols` function.**
+  The `symbols()` function supports creating one or multiple symbols at a
+  time:
```
```suggestion
  The `symbols()` function supports creating one or more symbols at once:
```

smichr commented on doc/src/explanation/best-practices.md:
```diff
@@ -0,0 +1,1329 @@
+
+
+One should also take care to avoid writing `integer/integer` where both
+integers are explicit integers. This is because Python will evaluate this to a floating-point value before SymPy is able to parse it.
```
```suggestion
integers are explicit integers. This is because Python will evaluate
this to a floating-point value before SymPy is able to parse it.
```

asmeurer replied:
I'm having difficulty reading the GitHub diff. Is there an actual change to the text here or is this just a line wrap?

moorepants commented on doc/src/explanation/best-practices.md:
```diff
@@ -0,0 +1,1329 @@
+  There also exist some alternatives to {func}`~.symbols`. Two alternatives
+  are using the {class}`~.Symbol` constructor directly and {mod}`sympy.abc`.
+  These are both fine, but they are less general than {func}`~.symbols`, so
+  just using `symbols()` is generally preferred. {class}`~.Symbol` only
```
Is there a reason `symbols()` is preferred? I recall seeing the use of `Symbol()` in user code discouraged over the years, but never knew why. This doesn't really explain why.

Is it that we want to promote using `symbols()` in our documentation just for style sake? After reading this I'm guessing that all three methods should likely be equally promoted.

oscarbenjamin replied:
There is another reason why many people will not want to use symbols which is that it cannot be typed properly because it returns different types based on the value of its argument:
```python
In [1]: type(symbols('x'))
Out[1]: sympy.core.symbol.Symbol

In [2]: type(symbols('x y'))
Out[2]: tuple
```
The `cls=` argument could be typed properly but distinguishing strings like "x" vs "x y" is not something that any type system can do.

Also `Symbol` is more efficient if you just want to make one symbol.

This is sort of mentioned below "if you want to be sure you are only creating one symbol" but it might be worth clarifying why you might want to ensure this: for example if the symbol name is a string that is passed through some end user API then the user might pass a string with spaces e.g. `Matrix([[1,2],[3,4]]).charpoly('long name')`

asmeurer replied:
Symbol is fine. I just think it's simpler to have one function that can create one symbol or multiple symbols instead of one function that creates one symbol and another function that creates multiple symbols. With `symbols()`, you basically only ever need the one function, which simplifies things.

sylee957 replied:
> you basically only ever need the one function, which simplifies things.

I just stopped using that in production code because `symbols` is unable to infer the types.
In the production code, you anyway are encouraged to do explicit imports from SymPy, so combined that you need to explicitly list you use in SymPy, I rather find that argument opinionated that that using one function having everything combined is really easier.

sylee957 replied:
On the other hand, I'd like to listen to the opinion about if `symbols` is used inside for-loop sounds awkward or should be best practice.

asmeurer replied:
> Also if assumptions are involved, symbols doesn't make things easier because it sets assumptions simultaneously. And you eventually need to write one symbol for each lines to give according assumptions

In that case you end up with something like what is used in isympy. Different lines for each type of symbol. In my experience, most people just set everything to be real or positive anyway.

```py
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
```

> On the other hand, I'd like to listen to the opinion about if symbols is used inside for-loop sounds awkward or should be best practice.

I'm not following what you are asking here. What do you mean by using symbols in a for loop?

sylee957 replied:
> I'm not following what you are asking here. What do you mean by using symbols in a for loop?

I would rather mean in other context, like building a web app where user can give any kind of equation. I don't use symbols because it parses the spacings.

I don't think that it's worth discussing whether to prefer symbols or symbol. 

In fact, the way how Symbol works behind is one of the easiest things that people can understand.
And it's better to let them trust their own sketches of how it works, than recommending something.

(I have other point of view that programming with type constructor as parameters like cls=Function is not becoming better nowadays because it limits type inference, but that's other topic)

asmeurer replied:
Before `symbols`, the common practice was to use `map(Symbol, 'xyz')`. But this is harder to read, and breaks down if you want symbol names with more than one character.

> In fact, the way how Symbol works behind is one of the easiest things that people can understand.

That's really not my experience, especially for people who are new to SymPy (and Python). The way symbols work in SymPy is surprising, especially for people coming from other systems where symbols are defined automatically. Even things like the fact that the variable name has no bearing on the symbol name throws a lot of people (people really have to work with Python for some time before they really *get* how variables work). 

I've seen people use things like `x = S('x')` to create a symbol, which is definitively **not** what you should be doing. Or people just parse full strings of expressions and try to do string expression manipulation (there are other sections in this guide about that). 

For these new users, `symbols` is the best choice.

I agree that if you have programmatic input, you should probably not be using `symbols`. That doesn't apply to 99% of SymPy users though.

Anyway, I'll probably rewrite this section a bit to be a little clearer on these points.

asmeurer replied:
I've rewritten this section.

sylee957 commented on doc/src/explanation/best-practices.md:
```diff
@@ -0,0 +1,1329 @@
+
+## Custom SymPy Objects
+
+SymPy is designed to be extended with custom classes, typically by subclassing
+{term}`Basic`, {term}`Expr`, or {term}`Function <Function (class)>`. All the
+symbolic classes in SymPy itself are written this way, and the points here
+apply equally to them as to user-defined classes.
+
+For an in-depth guide on how to write a `Function` subclass, see the [guide on
+writing custom functions](custom-functions).
```
My opinion is SymPy rarely need to be extended though.

For example, I'm not sure about what kind of things are more capable to do with custom functions with `eval`, than `Function` with substitution)

asmeurer replied:
Both this guide and the one linked here apply equally well to user code and library code. Every class in SymPy is a "custom object".

asmeurer replied:
Basically, I wrote this section of this guide and the custom functions guide primarily for for SymPy developers, because we were sorely lacking in that sort of documentation and there are a lot of things involved. But everything there is public API and completely supported for users. It's true that the vast majority of users shouldn't be doing anything this advanced, but there are some who do, and there are legitimate reasons to do so. We should always aim to make the methods used to extend SymPy internally to also be usable by end-users. That way we keep the APIs for doing so clean and sufficiently extensible.

sylee957 commented on doc/src/explanation/best-practices.md:
````diff
@@ -0,0 +1,1329 @@
+
+  ```
+  >>> from sympy import solve
+  >>> solve('x_{2} - 1') # doctest: +SKIP
````
I'm not sure if the LaTeX subscript is the best way to choose between `x_2, x_{2}, x2, ...` though
There is a problem that different expressions look same with same subscript, but compares differently `x_2, x_{2}, x2, ...`

asmeurer replied:
I agree that this is somewhat contrived. The point here is just to show how things can break, not to give a recommendation on how to name symbols. Anyway, I have seen users making symbol names like this.

oscarbenjamin commented on doc/src/explanation/best-practices.md:
```diff
@@ -0,0 +1,1329 @@
+  However, this sort of situation is rare.
+
+(best-practices-eq)=
+### Don't Overwrite `__eq__`
```
I don't know if we want to mention this here but related to "avoid extra attributes" and "don't override `__eq__`" is the question of what to do when you *do* need to do these things.

It is rare that these things are needed but there are examples e.g. `Poly`. Storing the `args` of a Poly as Expr would be very inefficient so it is not done and also the domain argument is not representable as `Basic` (possibly what this shows is that `Poly` should not be `Basic` at all). Likewise `ImmutableMatrix` has a `DomainMatrix` as its internal representation and it would be inefficient to always store the `args` in that case.

A more common situation where you do need to attach additional data not in the `args` is for atoms.

The question then is what you *should* do e.g. to implement an atom or some other Basic subclass whose constructor arguments will *not* end up in the args or at least will not be stored explicitly in `_args`.

1. Don't override `__eq__` or `__hash__`: use `_hashable_content` instead.
2. Override `__new__` but *don't* pass the args to `super().__new__`.
3. Implement `__getnewargs__` (`__getnewargs_ex__` if keyword arguments are needed) for pickling etc.
4. If not an `Atom` then override `.args` as a property that will generate args on demand (and ensure that the `func-args` invariant holds).

asmeurer replied:
These are somewhat advanced things. I'm not even sure myself what the proper way to do some of those things would be (is it ever actually a good idea to override `_hashable_content`?). In my experience, most people who are trying to override `__eq__` are completely misguided and shouldn't be doing it at all. They are just trying to do something that goes against the SymPy norms.

I did mention overriding `func` as an advanced escape hatch. Maybe I can briefly mention these other things, but if anything, I feel like this just gets to the core of why this is a bad idea: there are lot of things here and it's very easy to get it wrong.

asmeurer committed:
Fix typo that was causing doctest failures on CI

asmeurer committed:
Cleanups to the no string inputs section of the best practices doc

asmeurer committed:
Remove TODO

It is too hard to give a simple concrete example here, as the issue tends to
happen when things are intertwined in complex ways.

sylee957 commented:
> (possibly what this shows is that Poly should not be Basic at all).

What are the best practices for object-oriented programming used inside SymPy in general?
I just recognize that the document is absent about that part too.
And I was very reluctant about working anything outside of Basic, because there is no real rule about the best practices.

oscarbenjamin commented:
> What are the best practices for object-oriented programming used inside SymPy in general?

Mostly best practices are just general best practices for Python. Having a single god-superclass like Basic does tend to confuse people into thinking that everything needs to subclass it though even if it should really be a separate class or just an ordinary function or something.

I would find it difficult to write a best practices for object-oriented programming in relation to the SymPy codebase when this codebase has shown me so many examples of bad practice. It is now much clearer to me why inheritance is just generally bad and to be avoided for the most part for example but it is used everywhere in this codebase.

asmeurer commented:
I agree that Poly really should not be Basic. I think Basic was just added to it at some point just as a hack to make the interface work, but it wasn't really well thought out at the time.

asmeurer committed:
Suggested changes

sylee957 commented:
Basic had used to enforce some things like immutability.
I still think that Poly should be immutable at least, to make promises for users.

asmeurer commented:
And see for instance 548120b074d (later reverted by 75bbeabc858). Clearly not much thought was put into it. Unfortunately, I think the original change was squashed (31c9aa43ffc), so you'd have to dig into the old PRs to find the reason.

asmeurer commented:
Basic doesn't really enforce immutability. You can't actually enforce immutability in Python (at least not in pure Python). All it does is provide an interface (i.e., `.args`), and forces that one interface to be immutable. But Poly doesn't actually even use that interface. It just [fakes it by providing a `.args` that is computed from its actual interface](https://github.com/sympy/sympy/blob/f8bed8c4ee4de92ec2bb9f3076fb45ed407994db/sympy/polys/polytools.py#L213-L215), which are the low-level polys objects (`.rep`). Poly subclassing from Basic has absolutely this actual internal representation is immutable. It does so happen that they are, but that's just because Poly was also designed to be immutable.

And again, they aren't *really*, they just are by the top-level interface, but if you really want you can dig down and mutate things, because the very bottom representation is a list and lists are mutable:

```
>>> a = Poly(x)
>>> a.rep.rep[0] = ZZ(2)
>>> a
Poly(2*x, x, domain='ZZ')
```

Poly is immutable in the sense that the top-level interface doesn't mutate things, and if you try to actually mutate it anyway, it breaks it because it isn't expecting that. It's not immutable in the sense that it's impossible to mutate it, because that sense of immutable is simply not achievable in (pure) Python

sylee957 commented:
I think that specifically, making it hard and ugly to fiddle in a way to mutate is okay to call it immutable though, because it will be apparent in review.

sylee957 commented:
The other thing particularly nice about `Basic`, besides the immutability, 
is the unified terminology (`args` for example),
and people are enforced to think from a minimal cartesian product of the definition, to make a good definition about objects.

I think that these particular things were really nice, and I tended to adopt in many parts of the projects outside of SymPy as well, because it seemed like a good solution for trust and communication issues
Otherwise, we may start to see some people insisting things like:

```
class Foo:
    def __init__(self, ...):
        ...
        self.something = something
        ...
        self.something_1 = self.somehing + ...
        self._something = ...
        self._post_init()
    
    def bar(self, ...):
        ...
        self._something2 = ...
        ...
```

asmeurer committed:
Rewrite section recommending symbols()

Should now be clear that both symbols() and Symbol() are OK. Also indicate
that S() should not be used to create symbols.

asmeurer commented:
> The other thing particularly nice about Basic, besides the immutability,
is the unified terminology (args for example),

I agree. `args` is a really nice feature of SymPy, that you can just treat every object uniformly. A lot of libraries get this wrong (compare how the standard library `ast` module works, for instance). 

But the problem is that Poly isn't really built in this way. The Poly `args` are just faked. They don't actually represent how the Poly is built:

```py
>>> Poly(x + 1).args
(x + 1, x)
>>> type(Poly(x + 1).args[0])
<class 'sympy.core.add.Add'>
```

If you try to manipulate a Poly's args, that's a huge mistake. You are incurring a performance cost to convert the Poly into an Expr, and that Expr then no longer has any of the advantages of Poly. The actual representation of Poly is the `.rep`. If you want to dig down into it, you should use that. You can't really operate on it uniformly, because things depend quite a bit on the specific implementation details of the internal representation (and in practice, you really shouldn't use the rep, but rather use the top-level polynomial algorithms implemented as methods on Poly).

The Basic superclass of Poly was added to make something or other work, but I would argue that whatever it was that it was supposed to make work shouldn't actually be doing whatever it is doing that it wants Poly to look like Basic. 

To put it another way, Poly is a high performance low level computation object. High performance computation objects should never automatically do something that makes them unperformant. Whatever operation would do that should be an error, and such conversions should always be done explicitly.

sylee957 commented:
> Poly is a high performance low level computation object.

I don’t think that I agree on that point though. Poly is technically not low level, but the stuff at DMP, or much more lower is. 
However, I agree on the point that Poly should not be a major component to build Expr though, Because the expression with symbols, and utilities like expand, factor should do that.

asmeurer commented:
What is a high or low abstraction depends on where you are standing. And of course saying something is "low level" doesn't mean it is the lowest level. The lowest level abstraction is the integrated circuitry and transistors. But from the point of view of the average SymPy user, Poly is low level, because it's not what they want to be using.

sylee957 commented:
> But from the point of view of the average SymPy user, Poly is low level, because it's not what they want to be using.

Poly is an mathematical object. And its name, or its semantics, really seems like it should work like mathematical object, like other SymPy Expression.

I don’t really get what is high or low level in this regards. But if it is just a mathematical object, it should work like mathematics.
Why Poly cannot be multiplied by Expr, that looks like polynomial, for instance, (that may be inefficient) is a mathematically sound argument.

oscarbenjamin commented:
> Why Poly cannot be multiplied by Expr, that looks like polynomial, for instance, (that may be inefficient) is a mathematically sound argument.

It is possible:
```python
In [1]: Poly(x) * x
Out[1]: Poly(x**2, x, domain='ZZ')
```
I think that allowing that is okay but it does not mean that Poly needs to be Basic or Expr. `Poly.__mul__` can allow `Expr` without `Poly` itself being a subclass of `Expr` or `Basic`.

Regardless of whether Poly is "high level" or "low level" it should not be Basic. There are two reasons that it could make sense for it to be Basic:

- If it would be used in the args of a larger expression.
- If any of the Basic methods made sense for Poly. 

None of the Basic methods does really make sense for Poly so allowing them is just confusing and they end up all just doing strange things:
```python
In [6]: Poly((x - 1)**2).subs(x, y)
Out[6]: Poly(y**2 - 2*y + 1, y, domain='ZZ')

In [7]: Poly((x - 1)**2).subs(y, z)
Out[7]: 
 2          
x  - 2⋅x + 1

In [8]: Poly((x - 1)**2).subs(x, sqrt(2))
Out[8]: 3 - 2⋅√2
```
Should subs return Poly or Expr? It should be well defined but it isn't because the method doesn't really make sense for Poly.

I don't know of anywhere in SymPy that uses Poly in the args of a larger expression and there are other ways to handle that anyway like just using `Expr`:
```python
In [4]: apart(1/(x**5 - x + 1), full=True).args
Out[4]: 
⎛                       4        3        2                  ⎞
⎜                  320⋅a    400⋅a    500⋅a    625⋅a   256    ⎟
⎜                - ────── - ────── - ────── - ───── + ────   ⎟
⎜ 5                 2869     2869     2869    2869    2869   ⎟
⎜w  - w + 1, a ↦ ─────────────────────────────────────────, w⎟
⎝                                 -a + x                     ⎠

```
This is a situation where Poly could be used but is not because we can just use the expression `w**5 - w + 1`.

Previously Poly was Expr but I changed it to Basic. The fact that Poly was ever Expr or Basic at all though just comes from a misguided desire to make everything inherit the god class. This sort of thing is inevitable when OOP is used excessively. People get carried away wanting difference classes to inherit each other rather than thinking about what the actual code should be doing and what the real data structures should be.

sylee957 commented:
However, to be productive, we anyway need what is better alternative than Basic, than pointing out the problem of Basic.
I'd just take `dataclass` as the best alternative:

I'm not happy with any of the other object-oriented designs from students, that stepped out of Basic, anyway,
`FPgroup`, `DifferentialExtension`, galois resolvents, plot backends, ... all have some problems

https://github.com/sympy/sympy/blob/411eb29a4bd65588da868344392b38b484112dc1/sympy/integrals/risch.py#L120-L287

(I suspect some of them were written by some people subscribe this thread, 
so you may try to defend your designs anyway.)

They poorly document about the structure of the objects, have very many attributes, very long initializers,
or in other way, the documentation is too prose such that the code is only readable by someone who can follow both the mathmatical details and many contexts how it is used in SymPy,

For instance,
```
@dataclass
class Poly:
    rep: Final[DMP | SMP]
````

Because it have very clear meaning, very reusable design, have very thin initializer, should use few parameters,
most of them should be immutable unless there is clear reason to be mutable,
should use static typing unless there is a clear reason to have type dynamic.
And I may not be one of the people who says that objects that doesn't follow this, is only happy for the authors who write them, but unhappy for everyone who reads them.

oscarbenjamin commented:
It is a bit unclear to me whether this guide is aimed at users or at contributors. Most of the text seems more aimed at users. In that context it seems out of place to talk about OOP practices.

If this is aimed more at contributors then I would write a section about exception handling because that is something that is handled poorly within the codebase.

asmeurer commented:
It's aimed at users, which is why it's in the main part of the docs and not the contribution section. Some parts are aimed at advanced users, which spills over a little bit into the usage of SymPy in library code, like the bits about custom objects. But it definitely should only be about usage of SymPy specifically, not general Python best practices. 

The discussion in the more recent comments got a little off topic and should be moved to an issue. 

We could add a separate page in the contributing section of the docs for general Python coding principles that aren't specific to SymPy that should be applied to the SymPy codebase.

asmeurer committed:
Merge branch 'master' into best-practices

asmeurer commented:
Any opposition to getting this merged? I've merged master which should hopefully get the CI working.

sylee957 commented:
I think that we should review which sections are for users and which sections should be for developers.

I think that the sections about `simplify`, custom objects, may not be very relevant for end-users,
or it is too difficult material to read for SymPy users

https://github.com/sympy/sympy/blob/a9765f6a8415eb2d19d1dae9f86aaadfe58754a2/doc/src/explanation/best-practices.md#avoid-simplify
https://github.com/sympy/sympy/blob/a9765f6a8415eb2d19d1dae9f86aaadfe58754a2/doc/src/explanation/best-practices.md#custom-sympy-objects

asmeurer commented:
The idea was for the "basic usage" and "advanced usage" top-level sections to be that delineation. The right sidebar on the rendered page makes it easier to see what the sections are https://output.circle-artifacts.com/output/job/994da500-93e9-4e7b-b061-89b2675f6c6c/artifacts/0/doc/_build/html/explanation/best-practices.html

oscarbenjamin commented:
There is a lot of good advice here and I don't object to merging this but I had similar thoughts to @sylee957. 

Looking through some things are not relevant to SymPy users although the headline says that the document is intended to describe best practice for users. Most things in "advanced usage" are not relevant to users but also some things in the other parts don't seem relevant. The parts not relevant to users seem directed at contributors and would be better in a separate page.

Generally it seems like this should be split into two separate documents.

asmeurer commented:
> Generally it seems like this should be split into two separate documents.

That sounds fine. Let's do it now, though, so we don't have to break links.
